### ISSUE TITLE
Iceberg `register_table` fails on metadata files created by Spark

### ISSUE THREAD

**Issue by sweetpythoncode** [trinodb/trino#16363]
I want to migrate Hive tables to Iceberg and register it in Trino but got this
Exception: `io.trino.spi.TrinoException: Invalid metadata location: s3://test-bucket/test_iceberg/iceberg_db/db/table_name/metadata/v1.metadata.json`

Steps to reproduce
- Create an empty table through the spark with this config(With Iceberg Jar)
```
    "spark.jars.packages": "org.apache.iceberg:iceberg-spark-runtime-3.2_2.12:1.1.0",
    "spark.sql.extensions":"org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions",
    "spark.sql.catalog.dev":"org.apache.iceberg.spark.SparkCatalog",
    "spark.sql.catalog.dev.type":"hadoop"
```

- Run `system.add_files` with existing data inside the bucket
- Iceberg generates `v{n}.metadata.json` files inside the metadata folder
- Run `CALL system.register_table` inside Trino and got the exception
P.S. Based on connector logic [here](https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java#L650), its looks like Trino expect files in this format `00001-uuid.metadata.json` and failed on `v1.metadata.json`

It affects only cases when side catalog is not used(like metastore,DynamoDB, etc)

**Comment by krvikash** [comment]
Hi @sweetpythoncode, I think you might be looking for https://github.com/trinodb/trino/pull/13493. @ebyhr is working on migrating a hive table to Iceberg table PR.

**Comment by electrum** [comment]
Could you show the full stack trace for the failure? If you're running with the CLI, add `--debug` to see the stack trace.

**Comment by ebyhr** [comment]
I guess the file name comes from [HadoopTableOperations](https://github.com/apache/iceberg/blob/c07f2aa/core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java#L258-L264). Looking into the details.

**Comment by sweetpythoncode** [comment]
@ebyhr Yes its true

**Comment by georgewfisher** [comment]
It looks like the problem here is that Trino and Spark use different numbering schemes for the metadata files. This is a problem specific to table registration, for scenarios where you do not specify the metadata location.

The Trino logic expects metadata files to be numbered like this: 
`<version>-<uuid/any text>.metadata.json`
Example: `00000-8a355886-e0d0-4516-88b3-e84c1e256b18.metadata.json`

Logic here: https://github.com/trinodb/trino/blob/f088b730b6e29fb5d5be53dc09cd48a983dc7ee6/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java#L705

However, the Iceberg spec says differently, and this is what Spark uses when writing with the Iceberg jars:

> Each version of table metadata is stored in a metadata folder under the table’s base location using a file naming scheme that includes a version number, V: v<V>.metadata.json.

https://iceberg.apache.org/spec/#file-system-tables

Therefore, you see files called `v6207.metadata.json`, which cannot be read by Trino because there's no dash in the filename on which to perform the substring. You will then see the exception (fyi @electrum): 
```
io.trino.spi.TrinoException: Invalid metadata location: abfss://data@XXXXXXX.dfs.core.windows.net/XXXXX/metadata/v1.metadata.json
	at io.trino.plugin.iceberg.IcebergUtil.parseVersion(IcebergUtil.java:643)
	at io.trino.plugin.iceberg.procedure.RegisterTableProcedure.getLatestMetadataLocation(RegisterTableProcedure.java:185)
	at io.trino.plugin.iceberg.procedure.RegisterTableProcedure.lambda$getMetadataLocation$1(RegisterTableProcedure.java:172)
	at java.base/java.util.Optional.orElseGet(Unknown Source)
	at io.trino.plugin.iceberg.procedure.RegisterTableProcedure.getMetadataLocation(RegisterTableProcedure.java:172)
	at io.trino.plugin.iceberg.procedure.RegisterTableProcedure.doRegisterTable(RegisterTableProcedure.java:144)
	at io.trino.plugin.iceberg.procedure.RegisterTableProcedure.registerTable(RegisterTableProcedure.java:113)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(Unknown Source)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(Unknown Source)
	at io.trino.execution.CallTask.execute(CallTask.java:211)
	at io.trino.execution.CallTask.execute(CallTask.java:71)
	at io.trino.execution.DataDefinitionExecution.start(DataDefinitionExecution.java:145)
	at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:252)
	at io.trino.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:143)
	at io.trino.$gen.Trino_0_407_0_0_3_snapshot____20230329_220005_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

My questions are:
* If table registration only inspects the version files at registration time, how do external (Spark) updates to that table get consumed? Not at all? Should we add a synchronize registration procedure?
* Is looks like Trino has diverged from the Iceberg spec. Should I make a change to Trino to allow discovery of both metadata numbering formats?

**Comment by georgewfisher** [comment]
This issue profoundly confused me until I read this section in the Iceberg spec:
https://iceberg.apache.org/spec/#metastore-tables

> Metastore Tables
> The atomic swap needed to commit new versions of table metadata can be implemented by storing a pointer in a metastore or database that is updated with a check-and-put operation [1]. The check-and-put validates that the version of the table that a write is based on is still current and then makes the new metadata from the write the current version.

> Each version of table metadata is stored in a metadata folder under the table’s base location using a naming scheme that includes a version and UUID: `<V>-<random-uuid>.metadata.json`. To commit a new metadata version, V+1, the writer performs the following steps

So, this is why this issue only occurs when you do not use the Hive Metastore. Iceberg names the metadata version files differently for metastore mode vs file system mode.

If Trino were to scan these `v<version>.metadata.json` file system metadata files:
* The table registration should declare that it's a filesystem table
* The table should be read only

or

* The table metadata files should be renamed when HMS registered, as otherwise a file system writer might continue to write to the table.

I tried a bi-mode solution [which I have experimented with here](https://github.com/georgewfisher/trino/blob/f463becc80a254fa03dbc1be113b67e998411726/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java#L709). However, I think it's invalid because once a table is registered in the HMS it should become a non-filesystem table. But we do not know this for sure - Spark could continue to write using the filesystem which could create a split brain.
